src/context.rs

```diff
@@ -532,11 +478,495 @@ pub fn get_version_str() -> &'static str {
     &DC_VERSION_STR
 }
 
+/// Represents a file in the blob directory.
+///
+/// The object has a name, which will always be valid UTF-8.  Having a
+/// blob object does not imply the respective file exists, however
+/// when using one of the `create*()` methods a unique file is
+/// created.
+#[derive(Debug)]
+pub struct BlobObject<'c> {
+    blobdir: &'c Path,
+    name: String,
+}
+
+impl<'c> BlobObject<'c> {
+    /// Creates a new blob object with a unique name.
+    ///
+    /// Creates a new file in the blob directory.  The name will be
+    /// derived from the platform-agnostic basename of the suggested
+    /// name, followed by a random number and followed by a possible
+    /// extension.  The `data` will be written into the file.
+    ///
+    /// # Errors
+    ///
+    /// [BlobErrorKind::CreateFailure] is used when the file could not
+    /// be created.  You can expect [BlobError.cause] to contain an
+    /// underlying error.
+    ///
+    /// [BlobErrorKind::WriteFailure] is used when the file could not
+    /// be written to.  You can expect [BlobError.cause] to contain an
+    /// underlying error.
+    pub fn create(
+        context: &'c Context,
+        suggested_name: impl AsRef<str>,
+        data: &[u8],
+    ) -> std::result::Result<BlobObject<'c>, BlobError> {
+        let blobdir = context.get_blobdir();
+        let (stem, ext) = BlobObject::sanitise_name(suggested_name.as_ref().to_string());
+        let mut name = format!("{}{}", stem, ext);
+        for attempt in 0..16 {
+            let path = blobdir.join(&name);
+            match fs::OpenOptions::new()
+                .create_new(true)
+                .write(true)
+                .open(&path)
+            {
+                Ok(mut file) => {
+                    file.write_all(data)
+                        .map_err(|err| BlobError::new_write_failure(blobdir, &name, err))?;
+                    let blob = BlobObject {
+                        blobdir,
+                        name: format!("$BLOBDIR/{}", name),
+                    };
+                    context.call_cb(Event::NewBlobFile(blob.as_name().to_string()));
+                    return Ok(blob);
+                }
+                Err(err) => {
+                    if attempt == 16 {
```

Comment:
what magical number is `16`?

Reply:
It's the last attempt in the for loop above.  The algorithm of trying N times with a new random number and than failing I left unchanged from before, but I thought it'd be good to expose the actual error rather on the last attempt rather than hide it.

I'll use a constant for the number.